core/src/smartcontracts/isi/world.rs

```diff
@@ -171,22 +171,16 @@ pub mod isi {
                         error!(%role_id, "role not found - this is a bug");
                     }
 
-                    Ok(AccountEvent::PermissionRemoved(account_id))
+                    Ok(AccountEvent::RoleRevoked(account_id))
                 })?;
             }
 
             wsv.modify_world(|world| {
-                for mut domain in world.domains.iter_mut() {
-                    for account in domain.accounts_mut() {
-                        account.remove_role(&role_id);
-                    }
-                }
-
                 if world.roles.remove(&role_id).is_none() {
-                    return Ok(RoleEvent::Deleted(role_id).into());
+                    return Err(FindError::Role(role_id).into());
                 }
 
-                Err(FindError::Role(role_id).into())
+                Ok(RoleEvent::Deleted(role_id).into())
             })
         }
     }
```

Comment:
Might be refactored as follows:
```rust
    impl<W: WorldTrait> Execute<W> for Unregister<Role> {
        type Error = Error;

        #[metrics("unregister_role")]
        fn execute(
            self,
            _authority: <Account as Identifiable>::Id,
            wsv: &WorldStateView<W>,
        ) -> Result<(), Self::Error> {
            let role_id = self.object_id;

            wsv.modify_world(|world| {
                if world.roles.remove(&role_id).is_none() {
                    return Err(FindError::Role(role_id.clone()).into());
                }
                Ok(RoleEvent::Deleted(role_id.clone()).into())
            })?;

            for domain in wsv.domains() {
                for account in domain.accounts() {
                    if account.contains_role(&role_id) {
                        wsv.modify_account(account.id(), |account| {
                            account.remove_role(&role_id);
                            Ok(AccountEvent::RoleRevoked(account.id().clone()))
                        })?;
                    }
                }
            }

            Ok(())
        }
    }
```

Reply:
This suggestion is valid as far as the chain does not already have that inconsistency (phantom roles attached to accounts). So I think we have 2 ways:
- Adopt this suggestion as is and encourage people to resolve existing inconsistencies by revoking phantom roles for each account. This option promises to return early from invalid unregistration calls
- Reverse the 2 blocks in the suggestion -- world modification and accounts modification. This option promises to wipe out if any phantom roles brought in the future

Reply:
Also i believe that such implementation has some kind of lock, it may happen because we are trying to modify account in closure while outer scope has reference to it.
Maybe it explains why we have `accounts_with_role` in the first place.

Reply:
It's possible to rewrite account role removal part in a more functional style, but I don't know if it's worth it.

```rust
wsv.domains()
    .iter()
    .flat_map(|domain| {
        domain
            .accounts()
            .filter_map(|account| {
                if account.contains_role(&role_id) {
                    return Some(account.id().clone());
                }
                None
            })
            .collect::<Vec<_>>()
    })
    .try_for_each(|account_id| {
        wsv.modify_account(&account_id.clone(), |account| {
            if !account.remove_role(&role_id) {
                error!(%role_id, "role not found - this is a bug");
            }

            Ok(AccountEvent::RoleRevoked(account_id))
        })
    })?;
```

Reply:
I don't think we need to. You can resolve this conversation